projects/validation-messages/src/lib/components/validation-messages/validation-messages.component.ts

```diff
@@ -109,107 +99,89 @@ export class ValidationMessagesComponent
     return this.host._control;
   }
 
-  observeInputValueChanges(): void {
-    if (!this.valueChanges) {
-      this.valueChanges = this.control.valueChanges
-        .pipe(takeUntil(this.ngUnsubscribe))
-        .subscribe(this.unsubscribeAndClearValueChanges);
+  get isControlDirtyOrTouched() {
+    return this.control && (this.control.dirty || this.control.touched);
+  }
 
-      setTimeout(() => {
-        this.cd.markForCheck();
-      });
-    }
+  ngOnInit(): void {
+    this._readFormControlByControlName();
+  }
+
+  ngAfterContentInit(): void {
+    this._readFormControlFromHost();
+    this._listenToFormChanges();
   }
 
-  parseApiErrorMessages(apiErrorMessages: ApiErrorMessages): void {
+  private _parseApiErrorMessages(apiErrorMessages: ApiErrorMessages): string[] {
     if (!apiErrorMessages) {
-      this.parsedApiErrorMessages = [];
-      return;
+      return [];
     }
 
     const messages =
       apiErrorMessages instanceof Array
         ? [...apiErrorMessages]
         : [apiErrorMessages];
-    this.parsedApiErrorMessages = messages.map(
-      (message: ApiErrorMessage | string) =>
-        message instanceof Object
-          ? this.validationMessagesService.parseApiErrorMessage(
-              message.message,
-              message.property
-            )
-          : message
-    );
-  }
-
-  ngOnInit(): void {
-    this.readFormControlByControlName();
-  }
-
-  ngAfterContentInit(): void {
-    this.readFormControlFromHost();
-
-    this.control.valueChanges
-      .pipe(takeUntil(this.ngUnsubscribe))
-      .subscribe(() => this.cd.markForCheck());
-  }
-
-  ngOnDestroy(): void {
-    this.ngUnsubscribe.next();
-    this.ngUnsubscribe.complete();
-  }
 
-  ngDoCheck(): void {
-    if (
-      this.control &&
-      ((this.control.invalid && this.control.touched) ||
-        (!this.control.invalid && this.shownErrors.length > 0))
-    ) {
-      this.updateErrorMessages();
-      this.cd.markForCheck();
-    }
+    return messages.map((message: ApiErrorMessage | string) =>
+      message instanceof Object
+        ? this.validationMessagesService.parseApiErrorMessage(
+            message.message,
+            message.property
+          )
+        : message
+    );
   }
 
-  private updateErrorMessages(): void {
-    this.shownErrors = [];
+  private _defineErrorsToBeShown() {
```

Comment:
I'm pretty sure we can refactor this function.
```ts
const errorMessages = Object.entries(this.control.errors)
  .filter(([propertyName, value]) => value && propertyName !== 'server')
  .map(([propertyName, value]) => this.validationMessagesService.getValidatorErrorMessage(
    propertyName,
    value,
    this.errorsMessages
  ));

if (!this._multiple() && errorMessages.length) return [errorMessages[0]];

return errorMessages || [];
```

let me know if it works and looks good to you

Reply:
yeah it it works and looks good to me